### PR TITLE
Feature/component/rich select

### PR DESCRIPTION
**Pull request by pppemo**
@ugogo I've only got the Readme pending, but it's now good for review

### CODE REVIEW

File: packages/css/src/components/Item.css [hunk 1]
```diff
@@ -21,7 +21,7 @@
   cursor: pointer;
 }
 
-.f-Item.has-action:hover {
+.f-Item.has-action:hover, .f-Item.is-focused {
```
Comment (ugogo) [inline packages/css/src/components/Item.css]:
```suggestion
.f-Item.has-action:hover, .f-Item.has-action:focus {
```
?

File: packages/react/src/components/Dropdown/index.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
+import React, { useState } from 'react';
+import PropTypes from 'prop-types';
+import cx from 'classnames';
+import ReactSelect from 'react-select';
+import Icon from '../Icon';
+import UiText from '../UiText';
+import { safeInvoke } from '../../utils';
+
+const customStyles = {
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.js]: Could you add those style via CSS instead of inline ones?

File: packages/react/src/components/Dropdown/index.js [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+  control: () => ({
+    border: 0,
+    backgroundColor: 'transparent',
+    minHeight: 'initial',
+    display: 'flex',
+    justifyContent: 'space-between',
+  }),
+  indicatorSeparator: provided => ({
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.js]:
```suggestion
  indicatorSeparator: styles => ({
```

File: packages/react/src/components/Dropdown/index.js [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    ...provided,
+    display: 'none',
+  }),
+  indicatorContainer: provided => ({
+    ...provided,
+    marginRight: '1rem',
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.js]: Use :flamingo: vars?

File: packages/react/src/components/Dropdown/index.js [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    ref={innerRef}
+    {...innerProps}
+    icon={Icon.ICONS.IconChevronUpDown}
+    className='f-FormEl-icon'
+    size={Icon.SIZES.L}
+  />
+);
+
+const CustomSingleValue = ({ children }) => (
+  <UiText as='span' type={UiText.VARIANTS.content}>
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.js]: `Item`?

File: packages/react/src/components/Dropdown/index.js [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    {children}
+  </UiText>
+);
+
+const buildCustomOption = (onChange, isSelectable) => ({
+  data,
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.js]: `data: { labelComponent }` maybe?

File: packages/react/src/components/Dropdown/index.js [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    focused: isFocused,
+    onClick: e => {
+      safeInvoke(onChange, data.value);
+      safeInvoke(data.labelComponent.props.onClick, e);
+      isSelectable ? innerProps.onClick(e) : setValue(null);
+    },
+  },
+});
+
+const parseListItemObjectsIntoOptionObjects = objects =>
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.js]: I don't get what this fn is doing and it's quite verbose 😅

File: packages/react/src/components/Dropdown/index.js [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+  [].concat(objects).map(object => ({
+    label: object.props.children,
+    value: object.props.optionValue || object.props.children,
+    labelComponent: object,
+  }));
+
+const Dropdown = React.forwardRef(
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.js]: 👍

File: packages/react/src/components/Dropdown/index.js [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+  (
+    { className, onChange, children, placeholder, isSelectable, ...props },
+    ref,
+  ) => {
+    const [hasFocus, setHasFocus] = useState(false);
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.js]: Is this required?

File: packages/react/src/components/Dropdown/index.js [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+const parseListItemObjectsIntoOptionObjects = objects =>
+  [].concat(objects).map(object => ({
+    label: object.props.children,
+    value: object.props.optionValue || object.props.children,
+    labelComponent: object,
+  }));
+
+const Dropdown = React.forwardRef(
+  (
+    { className, onChange, children, placeholder, isSelectable, ...props },
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.js]: Can an option not be selectable?

File: packages/react/src/components/Dropdown/index.js [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+    ref,
+  ) => {
+    const [hasFocus, setHasFocus] = useState(false);
+
+    return (
+      <div className='f-FormEl-wrapper f-Select-wrapper'>
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.js]: `f-Select-wrapper` is not used anywhere
If you can `Select-wrapper` too in the `Select` component it'd be great, thanks!

File: packages/react/src/components/Dropdown/index.js [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+        <ReactSelect
+          placeholder={placeholder}
+          options={parseListItemObjectsIntoOptionObjects(children)}
+          components={{
+            DropdownIndicator: CustomDropdownIndicator,
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.js]:
```suggestion
            DropdownIndicator,
```

File: packages/react/src/components/Dropdown/index.js [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+            Option: buildCustomOption(onChange, isSelectable),
+            Placeholder: CustomSingleValue,
+            SingleValue: CustomSingleValue,
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.js]: Would it be possible to have
```
Placeholder,
SingleValue,
```
?

File: packages/react/src/components/Dropdown/index.js [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+          }}
+          onFocus={() => setHasFocus(true)}
+          onBlur={() => setHasFocus(false)}
+          className={cx(
+            'f-FormEl',
+            'f-FormEl--withIcon',
+            'f-Select',
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.js]: I don't think `f-Select` is actually used somewhere

File: packages/react/src/components/Dropdown/index.js [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+            'f-Dropdown',
+            className,
+            { 'is-focus': hasFocus },
+          )}
+          isSearchable={false}
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.js]: This should be moved into a prop

```suggestion
          isSearchable={isSearchable}
```

File: packages/react/src/components/Dropdown/index.js [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+          styles={customStyles}
+          classNamePrefix='FlamingoDropdown'
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.js]:
```suggestion
          classNamePrefix='f-Dropdown-'
```

File: packages/react/src/components/Dropdown/index.js [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+          ref={ref}
+          {...props}
+        />
+      </div>
+    );
+  },
+);
+
+Dropdown.displayName = 'Dropdown';
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.js]: 👌

File: packages/react/src/components/Dropdown/index.js [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+
+Dropdown.propTypes = {
+  /** Make this true if you want this dropdown to behave like a select component */
+  isSelectable: PropTypes.bool,
+  className: PropTypes.string,
+  placeholder: PropTypes.string,
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.js]: I'm pretty sure you have some eslint errors with this, could you confirm?

File: packages/react/src/components/Dropdown/index.stories.js [hunk 19]
```diff
@@ -0,0 +1,44 @@
+import React from 'react';
+import { storiesOf } from '@storybook/react';
+import { withKnobs } from '@storybook/addon-knobs';
+import { action } from '@storybook/addon-actions';
+import { Dropdown } from '..';
+import { Item } from '..';
+
+const defaultProps = () => ({
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.stories.js]: Yeah you don't need this, you can just add an `id` for each instance
And the action should exist in the `Playground` story

File: packages/react/src/components/Dropdown/index.stories.js [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+  id: `dropdown-${Math.random()}`,
+  onChange: action('onChange'),
+});
+
+storiesOf('Dropdowns/Dropdown', module)
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.stories.js]:
```suggestion
const stories = storiesOf('Dropdown', module)
```

Could you follow the same pattern as for every story?

- Use `const stories = `
- Add Heading
- Have a Playground story with as many knobs as possible as long as they make sense

File: packages/react/src/components/Dropdown/index.stories.js [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+      <Item
+        type={Item.SIZES.NORMAL}
+        value='Take me there'
+        helper='Check what can really happend!'
+        onClick={action('onClick')}
+      >
+        Cars choice
+      </Item>
+      <Item
+        type={Item.SIZES.NORMAL}
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.stories.js]: I'm pretty sure this is the default prop

File: packages/react/src/components/Item/index.js [hunk 22]
```diff
@@ -35,7 +36,7 @@ const Item = React.forwardRef(
     const isMini = size === SIZES.MINI;
 
     const containerProps = onClick && {
-      onClick: safeInvoke(onClick),
+      onClick,
```
Comment (ugogo) [inline packages/react/src/components/Item/index.js]: Why did you remove the `safeInvoke` fn?

File: packages/css/src/components/Item.css (hunk 1 above)
Comment (pppemo) [reply packages/css/src/components/Item.css]: Can't be, this is for the sake of `react-select` and it's executing focus on element by adding `is-focused`

File: packages/react/src/components/Dropdown/index.js (hunk 2 above)
Comment (pppemo) [reply packages/react/src/components/Dropdown/index.js]: I was considering it, but then it would require each attribute to be followed by `!important` to override original style. Do you think that would be better?

File: packages/react/src/components/Dropdown/index.js (hunk 3 above)
Comment (pppemo) [reply packages/react/src/components/Dropdown/index.js]: 👍

File: packages/react/src/components/Dropdown/index.js (hunk 5 above)
Comment (pppemo) [reply packages/react/src/components/Dropdown/index.js]: WDYM?

File: packages/react/src/components/Dropdown/index.js (hunk 6 above)
Comment (pppemo) [reply packages/react/src/components/Dropdown/index.js]: 👍

File: packages/react/src/components/Dropdown/index.js (hunk 4 above)
Comment (pppemo) [reply packages/react/src/components/Dropdown/index.js]: That's another problem since it's the inline style, flamingo vars are in CSS variables, so cannot be used here.
We could use flamingo vars, if these all were in CSS with `!important`.
Both solutions stink a bit. Dunno. What do you thinkk?

File: packages/react/src/components/Dropdown/index.js (hunk 5 above)
Comment (pppemo) [reply packages/react/src/components/Dropdown/index.js]: That is `UiText` because it represents the _selected_ value in the viewport of the select. Per design, even if the option itself is `Item` - meaning complex and rich, with arrows etc, only test is shown. This may be discussed with Juan further on if he thinks it's ok

File: packages/react/src/components/Dropdown/index.js (hunk 7 above)
Comment (pppemo) [reply packages/react/src/components/Dropdown/index.js]: Now that I've read it, yeah you're right, it's not that straightforward :)

So what it does:
1. What you're passing to `Dropdown` is an array of `Item`s
2. What you need to pass to `react-select` is a JSON label-value array.
3. So what it does it's wrapping every single `Item` into label-value object

Agree name is verbose. Can't think of anything better though. Have any ideas?

File: packages/react/src/components/Dropdown/index.js (hunk 9 above)
Comment (pppemo) [reply packages/react/src/components/Dropdown/index.js]: yes, b/c `react-select` doesn't have `:focus` selector itself. So if I want to change a style based on focus, I need to use `onFocus` and `onBlur`, store state, and apply `is-focused` style. See your first comment

File: packages/react/src/components/Dropdown/index.js (hunk 10 above)
Comment (pppemo) [reply packages/react/src/components/Dropdown/index.js]: That's for the case when the `Dropdown` will work as a menu (that was one of the requirements from Juan). Then once you click an option, I believe it shouldn't become "selected". You can click it, action will happen, but the value will not jump in the viewport of the component.

File: packages/react/src/components/Dropdown/index.js (hunk 11 above)
Comment (pppemo) [reply packages/react/src/components/Dropdown/index.js]: Done in both components!

File: packages/react/src/components/Dropdown/index.js (hunk 12 above)
Comment (pppemo) [reply packages/react/src/components/Dropdown/index.js]: 👍

File: packages/react/src/components/Dropdown/index.js (hunk 13 above)
Comment (pppemo) [reply packages/react/src/components/Dropdown/index.js]: I've changed `SingleValue`, but it also goes to `Placeholder` so L91 has to remain as is

File: packages/react/src/components/Dropdown/index.js (hunk 14 above)
Comment (pppemo) [reply packages/react/src/components/Dropdown/index.js]: Gone!

File: packages/react/src/components/Dropdown/index.js (hunk 15 above)
Comment (pppemo) [reply packages/react/src/components/Dropdown/index.js]: I'd leave it as is. Searching is not in the scope, and enabling this would require implementation of searching entire content of `Item` component in order to make it work intuitively.

File: packages/react/src/components/Dropdown/index.js (hunk 16 above)
Comment (pppemo) [reply packages/react/src/components/Dropdown/index.js]: That's actually not used anymore. Removed the line.

File: packages/react/src/components/Dropdown/index.js (hunk 18 above)
Comment (pppemo) [reply packages/react/src/components/Dropdown/index.js]: And you were pretty right! Made it `isRequired`

File: packages/react/src/components/Dropdown/index.stories.js (hunk 19 above)
Comment (pppemo) [reply packages/react/src/components/Dropdown/index.stories.js]: 👍

File: packages/react/src/components/Dropdown/index.stories.js (hunk 20 above)
Comment (pppemo) [reply packages/react/src/components/Dropdown/index.stories.js]: Done, yeah sorry didn't notice the patter.
I think it would be possible to implement a decorator to do it for us. Will add it to my TODO and take care of it when have time :)

File: packages/react/src/components/Dropdown/index.stories.js (hunk 21 above)
Comment (pppemo) [reply packages/react/src/components/Dropdown/index.stories.js]: 👍

File: packages/react/src/components/Item/index.js (hunk 22 above)
Comment (pppemo) [reply packages/react/src/components/Item/index.js]: I didn't have an understanding why, but this line was crashing when `Item` was used with `Dropdown` with `react-select` :/
However, existence of `onClick` is checked anyways in L37 right?

File: packages/react/src/components/Dropdown/index.js (hunk 4 above)
Comment (ugogo) [reply packages/react/src/components/Dropdown/index.js]: You should still be able to do `marginRight: 'var(...)',` 🤔

File: packages/react/src/components/Dropdown/index.js (hunk 7 above)
Comment (ugogo) [reply packages/react/src/components/Dropdown/index.js]: `getItems`?

File: packages/react/src/components/Dropdown/index.js (hunk 10 above)
Comment (ugogo) [reply packages/react/src/components/Dropdown/index.js]: 👍

File: packages/react/src/components/Dropdown/index.js (hunk 15 above)
Comment (ugogo) [reply packages/react/src/components/Dropdown/index.js]: 👍

File: packages/react/src/components/Item/index.js (hunk 22 above)
Comment (ugogo) [reply packages/react/src/components/Item/index.js]: Yep good catch 👍

File: packages/react/src/components/Dropdown/index.stories.js [hunk 23]
```diff
@@ -1,44 +1,100 @@
 import React from 'react';
 import { storiesOf } from '@storybook/react';
-import { withKnobs } from '@storybook/addon-knobs';
 import { action } from '@storybook/addon-actions';
 import { Dropdown } from '..';
 import { Item } from '..';
+import { boolean, select, text, number } from '@storybook/addon-knobs';
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.stories.js]: Please have a look at your eslint errors...

File: packages/react/src/components/Dropdown/index.stories.js [hunk 24]
```diff
@@ -1,44 +1,100 @@
 import React from 'react';
 import { storiesOf } from '@storybook/react';
-import { withKnobs } from '@storybook/addon-knobs';
 import { action } from '@storybook/addon-actions';
 import { Dropdown } from '..';
 import { Item } from '..';
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.stories.js]: Would be nice to avoid importing from the `components/index` file as we might end with a circular dependencies
```
import Item from '../Item';
```

File: packages/css/src/components/Dropdown.css [hunk 25]
```diff
@@ -0,0 +1,3 @@
+.f-Dropdown {
+  padding: 0 !important;
```
Comment (ugogo) [inline packages/css/src/components/Dropdown.css]: Why `!important`?

File: packages/react/src/components/Dropdown/index.js (hunk 2 above)
Comment (ugogo) [reply packages/react/src/components/Dropdown/index.js]: Can it be achieved by increasing the CSS selector level?
`.f-Dropdown .foo.bar`?

File: packages/react/src/components/Dropdown/index.js (hunk 5 above)
Comment (ugogo) [reply packages/react/src/components/Dropdown/index.js]: 👍

File: packages/react/src/components/Dropdown/index.js [hunk 26]
```diff
@@ -0,0 +1,128 @@
...
+  />
+);
+
+const SingleValue = ({ children }) => (
+  <UiText as='span' type={UiText.VARIANTS.content}>
+    {children}
+  </UiText>
+);
+
+const buildCustomOption = (onChange, isSelectable) => ({
```
Comment (ugogo) [inline packages/react/src/components/Dropdown/index.js]:
```suggestion
const buildOpt = (onChange, isSelectable) => ({
```
?
